### PR TITLE
WooCommerce: Add edit actions for Attributes

### PR DESCRIPTION
**Pull request by coderkevin**
Product Variations are extra complex, so this adds some special actions to utilize the rest of the Product Edit actions/reducer in #12921.

To Test:
1. `make test` and ensure all tests pass.

### CODE REVIEW

**Comment by allendav** [comment]
The `WOOCOMMERCE_EDIT_PRODUCT_VARIATION_TYPE` action and reducer could almost be used to edit product attributes in general.  Would you consider a more generic action and reducer that could do attributes and have the variation property included in the payload? That way you'd get not just the ability to edit variations, but non-variation attributes too for just a little bit more :)

**Comment by coderkevin** [comment]
@allendav, that's a good point. @justinshreve and I agree, so we made this apply to attributes. The variation-specific stuff really needs to be in selectors mostly, and the component code should be able to handle the rest (like setting `variation: true`, etc.)

**Comment by coderkevin** [comment]
@justinshreve also looked at the core WC code and found that attribute index order matters, so we'll be updating this code to use unique ids in the edit state, like you and I had discussed earlier.

File: client/extensions/woocommerce/state/ui/products/edits-reducer.js [hunk 1]
```diff
@@ -71,14 +71,26 @@ function editProduct( array, product, data ) {
 	return _array;
 }
 
-function editProductAttribute( attributes, attributeIndex, data ) {
+function editProductAttribute( attributes, attribute, data ) {
 	const prevAttributes = attributes || [];
-	const index = ( isNumber( attributeIndex ) ? attributeIndex : prevAttributes.length );
+	const uid = attribute && attribute.uid || uniqueId( 'edit_' );
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/edits-reducer.js]: Nice :)

**Comment by justinshreve** [comment]
@allendav I made a small change here if you want to take a quick look.

`uniqueId` by itself won't work here if we already have unsaved attributes in state from a previous session and go to add a new one. uniqueId basically just [increments a counter](https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L16143). That might not cause issues with the UI we are implementing since we would discard unsaved edits, but it still bit me since we are not doing that yet, and seems like it could be an issue if we ever decide to do things differently.

I'm appending a timestamp so these temporary IDs can still be unique, as well as keeping in `uniqueId` so two attributes added one after each other very quickly are still unique.

**Comment by allendav** [comment]
@justinshreve wrote:

> I'm appending a timestamp so these temporary IDs can still be unique, as well as keeping in uniqueId so two attributes added one after each other very quickly are still unique.

Does that (adding a timestamp) make the reducer impure?

**Comment by coderkevin** [comment]
> uniqueId basically just increments a counter.
>I'm appending a timestamp so these temporary IDs can still be unique

@justinshreve I'm not understanding why this should be an issue. Could you explain why this became a problem? Thanks.

**Comment by justinshreve** [comment]
I was having an issue where the first row would be blown away when coming back to the form and clicking Add new variation, because the ID would be edit_1 again. I was thinking that could be an issue if we ever wanted to hold multiple creates, but I realize now that would be under a separate product and shouldn't cause issues.

**Comment by justinshreve** [comment]
Though if we ever allowed the edits to be persisted instead of blown away if not saved wouldn't we have an issue?

**Comment by coderkevin** [comment]
The counter is based on a lodash module global. It shouldn't be resetting per-run. But I guess if the edit is persisted on the browser between refreshes, that could be a problem.

**Comment by coderkevin** [comment]
It looks like the edits were being persisted without a schema, which is a big no-no. I've got a fix for that here: https://github.com/Automattic/wp-calypso/pull/13320

**Pull request merged**